Allow truly asynchronous session negotiation without use of the base TCP class methods

dwmw2 opened the pull request:
- Promote NegotiateSession() to be a first-class citizen and public method of tuyaAPI.
- Remove the protocol-specific version of it which calls into the network send/receive functions.
- Let the protocol provide only BuildSessionMessage() and DecodeSessionMessage() calls, which a generic NegotiateSession can use.


This allows us to use the TuyaAPI protocol classes in environments like ESPHome where the socket handling needs to be done differently; you don't get to assume native BSD sockets. We can build a multi_kwh_meter example which doesn't use threads too, and just uses a poll() loop on its sockets.

dwmw2 committed:
Promote NegotiateSession to first-class tuyaAPI method

Add a NegotiateSession() method. For the protocols before 3.4 which don't
do actual session negotiation, this just sets the encryption key.

For the 3.4 protocol it does the full negotiation on the wire as it did
before, now under the explicit control of the user instead of implicitly
blocking in the first BuildTuyaMessage() call.

This change also removes the encryption_key parameter from BuildTuyaMessage()
and DecodeTuyaMessage() since the key is now stored in a member variable.

The examples now call NegotiateSession(device_key) after ConnectToDevice().

dwmw2 committed:
Move NegotiateSession() to base tuyaAPI class

This makes negotiation in the actual protocols properly non-blocking,
with BuildSessionMessage() and DecodeSessionMessage() methods. All
actual TCP communication is driven outside the protocol code.

NegotiateSession() is now just a simple loop which uses those functions
along with the blocking TCP send() and receive() functions — and like
those TCP functions, it's *optional*. A caller who wants properly
non-blocking operation can just call the protocols directly instead.

There are default BuildSessionMessage()/DecodeSessionMessage() methods
for the older protocols which don't do negotiation on the wire, so
the call to NegotiateSession() only sets the encryption key, as before.

dwmw2 commented:
Continuing the discussion from my first prototype of this in https://github.com/dwmw2/tuyapp/commit/bbb33b11afc7df35ed0fc83c4c247db4be256a54#r170838025

> Unsure. Still leaning towards having separate threads for each device as I don't believe that polling them in sequence will work with the pretty much complete randomness of the UPDATE_DPS call. Adding non common elements to every thread seems wasteful.

I'm not sure I see the issue with `UPDATE_DPS`. We just send it in place of a `HEART_BEAT`, don't we? Whenever we haven't received something from the device for a certain period of time? That's easy enough in the event-driven model, and the `HEART_BEAT` is already implemented in my ESPHome code. I need to add an option to do `UPDATE_DPS` instead. I've mostly "corrected" my Tuya power meter devices by installing ESPHome+LibreTiny on them, but I do have a couple of still-boxed spares here which I can test with.

> BTW I just committed an update to the TuyaTCP class. It's strictly non-blocking at the network level now. The receive() function now has a fourth optional parameter that allows the calling app to take control over reattempts and most importantly allow implementation of a break-out routine.

Nice. So once it's connected and the session established, an application can call into TuyaAPI only when that socket polls as readable anyway, or when it has decided the time has come to send a `HEART_BEAT`/`UPDATE_DPS`.

It's all very well having non-blocking I/O at the socket level; next we need to bring that ability to the tuyaAPI methods. You've handled `receive()`, but `ConnectToDevice()` is still a blocking API, even if it *internally* uses a non-blocking socket and then just loops waiting for it. But that's OK, the caller doesn't even have to use that.

The `NegotiateSession()` API is the more important one. Whatever it does internally, it doesn't return to the *caller* until it's waited for network I/O in a loop. And because it's called implicitly by `BuildTuyaPacket()`, that means `BuildTuyaPacket()` is a blocking API too.

This PR addresses that, by pulling out `NegotiateSession()` to be an explicit API (and cleaning up all the passing around of the device key a bit). So `BuildTuyaPacket()` is simple and guaranteed to be instantaneous again.

And then it cleans up `NegotiateSession()` to be a generic (and still blocking, but now optional) wrapper around the base protocol classes for `BuildSessionMessage()` and `DecodeSessionMessage()`.

Aside from the crypto parts to use mbedtls instead of OpenSSL (which PR #9 solves for me), that is all I need to have it working fully asynchronously in ESPHome.

dwmw2 commented:
It's optional (unless you want tuyapp to be usable in an environment like ESPHome), but here's what it looks like in a tuyapp example: https://github.com/gordonb3/tuyapp/commit/62dd75be19ae680551ac57f8f75ed74c40409320

Should be simple to make it send the UPDATE_DPS when needed and then it'd be a single-threaded asynchronous alternative to the multi-kwh-meter example.

gordonb3 commented:
> I'm not sure I see the issue with `UPDATE_DPS`. We just send it in place of a `HEART_BEAT`, don't we? Whenever we haven't received something from the device for a certain period of time? That's easy enough in the event-driven model, and the `HEART_BEAT` is already implemented in my ESPHome code. I need to add an option to do `UPDATE_DPS` instead. I've mostly "corrected" my Tuya power meter devices by installing ESPHome+LibreTiny on them, but I do have a couple of still-boxed spares here which I can test with.

As I've come to understand, UPDATEDPS is not considered a command, but more of an invitation to send a notification if something changes. If nothing changes, then nothing gets sent. I've also recently discovered that repeating the UPDATEDPS request if nothing was returned, will cause the device to hang up after a couple of times. You need to send HEARTBEAT and only repeat UPDATEDPS after you got a response from the previous one, even if that takes hours.
 
> The `NegotiateSession()` API is the more important one. Whatever it does internally, it doesn't return to the _caller_ until it's waited for network I/O in a loop. And because it's called implicitly by `BuildTuyaPacket()`, that means `BuildTuyaPacket()` is a blocking API too.
 
As I am reading, the NegotiateSession() is only called once, on the first message being send? It seems to me that making that blocking towards the caller is exactly the same reasoning why I made `ConnectToDevice()` blocking. Ideally it only gets executed once for each device and you can't do anything without it having completed (for that device). I'm starting to get your point about circling a pool of FDs from a single thread, though I believe this concept might prove to cause issues under Windows (think winsock doesn't allow reusing sockets, so eventually you'll run out of available ones?). Not a big issue as far as I'm concerned, to add an optional instant return flag to `ConnectToDevice()` as well, so that a single threaded caller can also decide what to do while waiting for EINPROGRESS to change.

dwmw2 commented:
> As I am reading, the NegotiateSession() is only called once, on the first message being send? It seems to me that making that blocking towards the caller is exactly the same reasoning why I made ConnectToDevice() blocking. Ideally it only gets executed once for each device and you can't do anything without it having completed (for that device).

Yep, that's a perfectly valid model for an application which is OK to block. If you take a look at the latest PR (#10)  we have tuyaTCP providing the synchronous methods just like before, and the only real difference is that there's an explicit NegotiateSession call to set the device key instead of passing it into every later message call.

And I *also* show how you can do a tuyaAsync which never ever blocks, even for its initial TCP connection and the negotiation. Which is how I use it from ESPHome.

> I'm starting to get your point about circling a pool of FDs from a single thread, though I believe this concept might prove to cause issues under Windows (think winsock doesn't allow reusing sockets, so eventually you'll run out of available ones?). 

It doesn't use any more sockets than before; one for each device it's talking to. It's just that it drives them in an entirely non-blocking way.

> Not a big issue as far as I'm concerned, to add an optional instant return flag to ConnectToDevice() as well, so that a single threaded caller can also decide what to do while waiting for EINPROGRESS to change.

Yep, that's what tuyaAsync does. Fits into a standard application event loop, like for example from a GTK+ application (or any windowing system really). 

We don't *have* to provide that tuyaAsync class; it's just an example of how it can be used from e.g. the [ESPHome component](https://github.com/dwmw2/esphome/commit/tuya_tcp). The key thing is that none of the core protocol-specific methods can ever block. Even during negotiation; they're just called to "give me the next packet to send" and "here's the next packet I received".

gordonb3 commented:
> It doesn't use any more sockets than before; one for each device it's talking to. It's just that it drives them in an entirely non-blocking way.

Misunderstood. The point is that the devices can reset the connection and you must then reconnect. Not real big on Windows, and judging your code you are even less interested in Windows, but as I was told it cannot reuse the socket within the same app/thread, so it must create a new one if you aim to keep the thread. May be less of a problem though, as with the latest changes I made to the `tuyaTCP` class (have not pushed all of it yet) I'm seeing far less resets.

> We don't _have_ to provide that tuyaAsync class; it's just an example of how it can be used from e.g. the [ESPHome component](https://github.com/dwmw2/esphome/commit/tuya_tcp). The key thing is that none of the core protocol-specific methods can ever block. Even during negotiation; they're just called to "give me the next packet to send" and "here's the next packet I received".

Not seeing the `tuyaAsync` class in this PR, but I believe we can offer both modes from the same class, possibly use a #define to reduce image size by skipping code that becomes orphaned when async mode is selected.

Which brings me to the point that I believe that this PR is moving towards a too strict single use case that is probably not the best to run on embedded. I think we should ditch the global variables, rather than add new ones, as to make the functions re-entrant.

dwmw2 commented:
Windows is no different to any other OS here. If the other end closes a socket, then yes that file descriptor isn't useful any more. You have to close it and make a new one. You can't just call `connect()` again on the old socket. That's true regardless of whether you are operating in blocking or non-blocking mode.

> Not seeing the tuyaAsync class in this PR, but I believe we can offer both modes from the same class, possibly use a #define to reduce image size by skipping code that becomes orphaned when async mode is selected.

Yes, that's exactly what it does. The core protocols now support completely non-blocking operation ("just give me the nett packet", etc.) while the tuyaTCP class is blocking as before but just happens now to be *optional*.

> Which brings me to the point that I believe that this PR is moving towards a too strict single use case that is probably not the best to run on embedded.

Hm? It's intended to move away from the single (blocking) use case in order to support *both* that, and the low-overhead embedded case.

>  I think we should ditch the global variables, rather than add new ones, as to make the functions re-entrant.

Did I add global variables? I didn't mean to; there was one global seqno in the initial 3.4 protocol but that was just a thinko and didn't I fix it already?

dwmw2 commented:
Here's an example of how the protocols can be used from a deeply embedded system using the platform's AsyncTCP and just building/deciding packets: https://github.com/dwmw2/esphome/blob/tuya_tcp/esphome/components/tuya_tcp/tuya_tcp.cpp

gordonb3 commented:
I have carefully reviewed this PR and must decline it.

Issue 1:
It was not clear why you made `encryption_key` a global variable. As I found during my review you chose to do so because as of API version 3.4 `BuildTuyaMessage()` and `DecodeTuyaMessage()` do not utilize `encryption_key` but a temporary issued session key instead. As such I understand your reasoning behind it. Changing the function definitions does however break backwards compatibility, which should not be done. Nor should existing implementations be made dependent on functions that effectively do nothing for the older supported version.

On a side note: I recklessly broke backwards compatibility myself on the `ConnectToDevice()` function, which I reverted earlier today. Truly sorry for that to those affected by it.

Issue 2:
`NegotiateSession()` relies internally on network feedback and therefore is not async as the PR implies to accomplish.

Additionally, I have noticed quite a lot of code replication with respect to `BuildTuyaMessage()` and `BuildSessionMessage()`, as well as their counterparts. I'm certain that this won't help achieving the goal to 'never block'. I have an idea how to fix this, which should also simplify the task to break up `NegotiateSession()` into actual async consumable elements.

dwmw2 commented:
>  certain that this won't help achieving the goal to 'never block'. I have an idea how to fix this, which should also simplify the task to break up NegotiateSession() into actual async consumable elements.

Yes, I've already done that as the next step.

dwmw2 commented:
The key isn't global, it's a member variable. I guess we could put it back into the message functions and the older protocols could use it from there. We could give it a default so it's optional.

Yes, NegotiateSession in tuyaTCP is intentionally blocking for ease of compatibility, while the actual protocol methods are non-blocking so callers who don't want to block don't have to.

dwmw2 commented:
> I have an idea how to fix this, which should also simplify the task to break up NegotiateSession() into actual async consumable elements.

You can see an example of how to use the actual async consumable elements which are already in this PR, at 
https://github.com/dwmw2/esphome/blob/tuya_tcp/esphome/components/tuya_tcp/tuya_tcp.cpp#L68

dwmw2 commented:
Now updated to preserve the existing `BuildTuyaMessage()` and `DecodeTuyaMessage()` APIs so that existing code which only supports older protocols can still just pass the device key in directly.

Obviously, NegotiateSession() still has to happen for newer protocols, and the blocking function remains just as it is today; just moved to the tuyaAPI class _(I'd like it to be in tuyaTCP and have that be a completely separate thing which HAS-A tuyaAPI, rather than a base class, but that can come later)_.

The important thing is that the blocking NegotiateSession() is now *optional*. It still exists for ease-of-use in cases where the caller is OK to block, but callers who need non-blocking behaviour can call the protocol `BuildSessionMessage()` and `DecodeSessionMessage()` directly instead.

And yes, there's probably some scope for reducing duplication between those and the general `BuildTuyaMessage()`/`DecodeTuyaMessage()` but that was also true before and it just a little but more obvious now as the code is being cleaned up, so that can be a later step. Trying to do things one step at a time.

dwmw2 commented:
Here's another example of the fully-asynchronous mode using "actual async consumable elements" that this PR already enables: a GTK+ app using GIO async socket operations: https://github.com/dwmw2/tuyapp/commit/3a5b828

dwmw2 committed:
Refactor tuyaAPI34 to eliminate code duplication

Extract common message-building code into BuildMessage34() helper function.
Both BuildTuyaMessage() and BuildSessionMessage() now use this helper,
eliminating ~40 lines of duplicated header/encryption/HMAC/trailer code.

dwmw2 commented:
The duplication between `BuildTuyaMessage()` and `BuildSessionMessage()` was already there before this PR, but it's certainly much more obvious now that this PR pulls the latter out to be an async consumable element in its own right. Added a commit on top to factor out a helper and avoid the duplication.

gordonb3 commented:
Please move this to /contrib/[yourproject] as it targets a too restricted usage scenario that falls outside the scope of this project.

dwmw2 commented:
No, this is just the basic requirement for asynchronous operation. If you don't want to support non-blocking operation, that's fine. I'll just keep a fork.
